sympy/vector/functions.py

```diff
@@ -196,24 +216,68 @@ def gradient(scalar, coord_sys):
     scalar : SymPy Expr
         The scalar field to compute the gradient of
 
-    coord_sys : CoordSysCartesian
-        The coordinate system to calculate the gradient in
-
     Examples
     ========
 
-    >>> from sympy.vector import CoordSysCartesian, gradient
-    >>> R = CoordSysCartesian('R')
+    >>> from sympy.vector import CoordSystem3D, gradient
+    >>> R = CoordSystem3D('R')
     >>> s1 = R.x*R.y*R.z
-    >>> gradient(s1, R)
+    >>> gradient(s1)
     R.y*R.z*R.i + R.x*R.z*R.j + R.x*R.y*R.k
     >>> s2 = 5*R.x**2*R.z
-    >>> gradient(s2, R)
+    >>> gradient(s2)
     10*R.x*R.z*R.i + 5*R.x**2*R.k
 
     """
+    # Check input
+    if isinstance(scalar, Zero) or scalar == 0:
+        return Vector.zero
+
+    if not issubclass(type(scalar), Expr) or scalar.is_Vector:
+        raise TypeError("Your input should be a scalar field.")
+
+    for atom in scalar.atoms():
+        if isinstance(atom, BaseScalar):
+            system = atom.system
+            break
```

Comment:
I hear you on that.  What we should focus on now is the original commits.  The removal of system parameter from *functions.py* should be seperate.  As @srjoglekar246 mentioned in a private message was that he designed CoordSysCartesian() to be a subclass of a general CoordSystem(R=3) class.  So maybe we should just keep that and have it inherited by the CoordSystem(R=3, sys=Cartesian) superclass if the user wants the system to be cartesian.  

What we need to decide on:
1. Do we want to restrict the CoordSystem() to Euclidian space? -- Depreciation process could be avoided in doing so. 
2. Have the superclass CoordSystem(R=3, sys=spherical) build the differential class within.

In future PR's:
1. Should we restrict vectors and scalar fields to one coordinate system at a time? -- Does Mathematica allow ```gradient(x*r)```?
2. Removal of the system requirement from *functions.py*
3. Vector functions and handling therein for *functions.py* -- ```type(f(C.x*C.i))``` gives ```f```
4. ...

Please note I'm just spitballing ideas here.  Feel free to add or tell me if it's a terrible idea.